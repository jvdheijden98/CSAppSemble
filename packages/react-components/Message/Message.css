.root {
  bottom: 10px;
  box-sizing: border-box;
  margin-left: 10px;
  margin-right: 10px;
  position: fixed;
  width: calc(100vw - 20px);
  z-index: 3000;
}

.messageEnter {
  opacity: 0;
}

.messageEnterActive {
  opacity: 1;
  transition: opacity 300ms ease-in;
}

.messageExit {
  opacity: 1;
}

.messageExitActive {
  opacity: 0;
  transition: opacity 300ms ease-in;
}

.content {
  min-height: 64px;
  position: relative;
}

.deleteButton {
  position: absolute;
  right: 5px;
  top: 5px;
}
