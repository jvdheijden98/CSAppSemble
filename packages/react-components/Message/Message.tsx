import { Message as IncomingMessage } from '@appsemble/types';
import classNames from 'classnames';
import * as React from 'react';
import { WrappedComponentProps } from 'react-intl';
import { CSSTransition, TransitionGroup } from 'react-transition-group';

import styles from './Message.css';
import msgs from './messages';

export interface UniqueMessage extends IncomingMessage {
  id: number;
}

export interface MessageProps extends WrappedComponentProps {
  messages: UniqueMessage[];
  remove: (message: UniqueMessage) => void;
}

export default class Message extends React.Component<MessageProps> {
  static defaultProps: Partial<MessageProps> = {
    messages: [],
  };

  componentDidUpdate(prevProps: MessageProps): void {
    const { messages, remove } = this.props;

    if (messages.length > prevProps.messages.length) {
      const message = messages[messages.length - 1];
      let { timeout } = message;

      if (!timeout && !message.dismissable) {
        timeout = 5e3;
      }

      if (timeout) {
        setTimeout(() => {
          remove(message);
        }, timeout);
      }
    }
  }

  onDismiss(message: UniqueMessage): void {
    const { remove } = this.props;

    remove(message);
  }

  render(): React.ReactNode {
    const { intl, messages } = this.props;

    return (
      <div className={styles.root}>
        <TransitionGroup>
          {messages.map(message => (
            <CSSTransition
              key={message.id}
              classNames={{
                enter: styles.messageEnter,
                enterActive: styles.messageEnterActive,
                exit: styles.messageExit,
                exitActive: styles.messageExitActive,
              }}
              timeout={300}
            >
              <article
                className={classNames(
                  'message',
                  message && message.color ? `is-${message.color}` : 'is-danger',
                )}
              >
                <div className={classNames('message-body', styles.content)}>
                  <span>{message && message.body}</span>
                  {message.dismissable && (
                    <button
                      aria-label={intl.formatMessage(msgs.dismiss)}
                      className={`delete ${styles.deleteButton}`}
                      onClick={() => this.onDismiss(message)}
                      type="button"
                    />
                  )}
                </div>
              </article>
            </CSSTransition>
          ))}
        </TransitionGroup>
      </div>
    );
  }
}
