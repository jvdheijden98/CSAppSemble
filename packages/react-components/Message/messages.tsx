import { defineMessages } from 'react-intl';

export default defineMessages({
  dismiss: 'Dismiss message',
});
