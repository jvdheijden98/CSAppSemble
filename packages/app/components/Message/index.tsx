import { Message } from '@appsemble/react-components';
import React from 'react';
import { connect } from 'react-redux';

import { State } from '../../actions';
import { remove } from '../../actions/message';

function mapStateToProps(state: State): Pick<React.ComponentProps<typeof Message>, 'messages'> {
  return {
    messages: state.message.queue,
  };
}

export default connect(mapStateToProps, { remove })(Message);
