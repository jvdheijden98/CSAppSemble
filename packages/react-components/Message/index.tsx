import { injectIntl } from 'react-intl';

import Message, { MessageProps, UniqueMessage } from './Message';

export { UniqueMessage };
export default injectIntl<'intl', MessageProps>(Message);
