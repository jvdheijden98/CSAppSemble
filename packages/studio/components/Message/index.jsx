import { Message } from '@appsemble/react-components';
import { connect } from 'react-redux';

import { remove } from '../../actions/message';

function mapStateToProps(state) {
  return {
    messages: state.message.queue,
  };
}

export default connect(mapStateToProps, { remove })(Message);
