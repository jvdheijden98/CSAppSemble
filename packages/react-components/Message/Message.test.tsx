import { shallow } from 'enzyme';
import React from 'react';

import Message, { UniqueMessage } from './Message';

describe('Message', () => {
  it('should render messages', () => {
    const messages: UniqueMessage[] = [
      { id: 1, body: 'Foo' },
      { id: 2, body: 'Bar', color: 'info' },
    ];
    const result = shallow(<Message messages={messages} remove={() => {}} />);

    expect(result).toMatchSnapshot();
  });

  it('should call remove after 5 seconds', () => {
    jest.useFakeTimers();

    const messages = [
      { id: 1, body: 'Foo' },
      { id: 2, body: 'Bar', color: 'info' },
    ];
    const mock = jest.fn();
    const result = shallow(<Message messages={[]} remove={mock} />);

    // Simulate adding messages over time
    result.setProps({ messages: [messages[0]] });
    jest.advanceTimersByTime(1000);
    result.setProps({ messages });

    expect(mock).not.toHaveBeenCalled();

    jest.advanceTimersByTime(4000);
    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenLastCalledWith(messages[0]);

    jest.advanceTimersByTime(1000);
    expect(mock).toHaveBeenCalledTimes(2);
    expect(mock).toHaveBeenLastCalledWith(messages[1]);
  });
});
