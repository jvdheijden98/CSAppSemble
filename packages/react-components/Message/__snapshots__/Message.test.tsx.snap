// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should render messages 1`] = `
<div
  className="root"
>
  <TransitionGroup
    childFactory={[Function]}
    component="div"
  >
    <CSSTransition
      classNames={
        Object {
          "enter": "messageEnter",
          "enterActive": "messageEnterActive",
          "exit": "messageExit",
          "exitActive": "messageExitActive",
        }
      }
      key="1"
      timeout={300}
    >
      <article
        className="message is-danger"
      >
        <div
          className="message-body content"
        >
          <span>
            Foo
          </span>
        </div>
      </article>
    </CSSTransition>
    <CSSTransition
      classNames={
        Object {
          "enter": "messageEnter",
          "enterActive": "messageEnterActive",
          "exit": "messageExit",
          "exitActive": "messageExitActive",
        }
      }
      key="2"
      timeout={300}
    >
      <article
        className="message is-info"
      >
        <div
          className="message-body content"
        >
          <span>
            Bar
          </span>
        </div>
      </article>
    </CSSTransition>
  </TransitionGroup>
</div>
`;
